src/ui/controllers/searchbariconcontroller.js

```diff
@@ -0,0 +1,83 @@
+import SearchStates from '../../core/storage/searchstates';
+import StorageKeys from '../../core/storage/storagekeys';
+import DOM from '../dom/dom';
+import LoadingIconState from '../statemachine/loadingiconstate';
+import SearchIconState from '../statemachine/searchiconstate';
+import DefaultIconState from '../statemachine/defaulticonstate';
+import { StateMachine } from '../statemachine/statemachine';
+
+/**
+ * Responsible for controlling the search bar icon behavior
+ */
+export default class SearchBarIconController {
+  constructor (config) {
+    this.iconIsFrozen = false;
+    this.inputEl = config.inputEl;
+    this.searchBarContainer = config.searchBarContainer;
+    this.onMouseUp = this.oneTimeMouseUpListener.bind(this);
+
+    const IconState = {
+      LOADING: new LoadingIconState(this),
+      SEARCH: new SearchIconState(this, config.useCustomIcon),
+      DEFAULT: new DefaultIconState(this, config.useCustomIcon)
+    };
+
+    const transitions = {
+      [IconState.LOADING.id]: {
+        SEARCH_COMPLETE: IconState.DEFAULT
```

Comment:
What do you think of having an enum for the 'SEARCH_COMPELTE', 'SUBMIT', 'FOCUS', and 'FOCUS_OUT' events?